keps/sig-api-machinery/4420-retry-generate-name/README.md

```diff
@@ -0,0 +1,566 @@
+suffix. We consider breaking thes projects disruptive enough that we would not
+want to do it immediately without ample warning and a long wait period for fixes
+to propagate to the ecosystem.
+
+This is why we favor this retry approach in the short term. It is an effective
+bandaid that makes the problem effectively go away for the vast majority of users
+and is something we can do now.
+
+Note that if we were to add more random suffix chars, we would like increase it
+to more than 11 to reduce the potential for collisions to probabilities similar
+to [git hash collision probabilities](https://git-scm.com/book/en/v2/Git-Tools-Revision-Selection#:~:text=The%20SHA%2D1%20digest,sand%20on%20the%20earth.).
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+- staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go
+  - Mock out name generator to use monotonically-increasing integers
+  - Verify that a resource is generated with the max retries when conflicting names
+    exist
+  - Verify that if max retries is exceeded that a already exists error is returned
+  - Verify that retries are not attempted when feature is disabled
+
+##### Integration tests
+
+<!--
+Integration tests are contained in k8s.io/kubernetes/test/integration.
+Integration tests allow control of the configuration parameters used to start the binaries under test.
+This is different from e2e tests which do not allow configuration of parameters.
+Doing this allows testing non-default options and multiple different and potentially conflicting command line options.
+-->
+
+<!--
+This question should be filled when targeting a release.
+For Alpha, describe what tests will be added to ensure proper quality of the enhancement.
+
+For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
+https://storage.googleapis.com/k8s-triage/index.html
+-->
+
+- Verify that retries are not attempted when feature is disabled
+- Verify that up to max retries are attempted when conflicting names exist
+- Verify that if max retries is exceeded that a already exists error is returned
+
+##### e2e tests
+
+This feature is not e2e testable. There is no predictable way to
+trigger name collisions.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Unit tests added. The `generic/registry/store_test.go` suite covers this functionality remarkably well.
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Add integration tests
+- Additional tests are in Testgrid and linked in KEP
+- Benchmark worst case: Max retries are attempted for all create requests.
+- Consider further optimizing. One possible option:  If the
+  first create attempt conflicts, do a loop to check etcd
```

Comment:
this could be ok after the first one.  I would like the first attempt to be "free" though.